fix: updated columns names

techbhavin opened the pull request:
## Description

> update column names in table widget 

> we have a key name as a number that is not valid as per javascript. example :  const test = { 1: “demo” }
> here `test.1` is invalid but `test["1"]` is valid and also we can change the key name from 1 to _1 ( valid key name )


Fixes #6671 

## Type of change

- Bugfix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes












## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>


    // Code coverage diff between base branch:release and head branch: fix/numeric-column-name 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.05 **(0.03)** | 36.49 **(0.03)** | 34.46 **(0.03)** | 55.57 **(0.02)**
 :green_circle: | app/client/src/utils/DSLMigrations.ts | 74 **(0.11)** | 64.99 **(-0.05)** | 70.21 **(0)** | 73.82 **(0.11)**
 :red_circle: | app/client/src/utils/migrations/TableWidget.ts | 80.38 **(-3.65)** | 73.91 **(-1.09)** | 85.29 **(-4.71)** | 81.28 **(-4.37)**
 :red_circle: | app/client/src/widgets/TableWidget/component/TableHelpers.ts | 96.88 **(-3.12)** | 64.29 **(-10.71)** | 83.33 **(-16.67)** | 96.15 **(-3.85)**
 :green_circle: | app/client/src/widgets/TableWidget/component/TableUtilities.tsx | 38.1 **(0.15)** | 23.89 **(0)** | 16.22 **(0)** | 36.36 **(0.18)**
 :green_circle: | app/client/src/widgets/TableWidget/widget/derived.js | 65.09 **(6.32)** | 61.54 **(2.26)** | 51.16 **(4.65)** | 63.08 **(4.32)**</details>

techbhavin commented:
/ok-to-test sha=d4ac084

riodeuno commented:
@techbhavin There isn't enough information in the PR about how we've solved the issue. 
Please answer the following queries in this and all future PRs, so that it is easier to review and understand.

- [ ] What is the strategy to solve this? Did we change the way we sanitize column keys?
- [ ] Will this change affect existing users? If yes, did we add a migration for this? If not, is it because it was never working for existing users?
- [ ] Which is the flow for which we have added new cypress tests?
- [ ] Do all new functions have unit tests associated with them?

techbhavin commented:
/ok-to-test sha=076a283

techbhavin commented:
> What is the strategy to solve this? Did we change the way we sanitize column keys?

the strategy behind this solution is to update numeric key to valid string key in data to access into autosuggestion
we handled that case into sanitizing column keys

> Will this change affect existing users? If yes, did we add a migration for this? If not, is it because it was never working for existing users?

yes, it will affect existing users which have numeric keys in data. we will add migration for that once we get approval for these changes 

> Which is the flow for which we have added new cypress tests?

we have added a jest case for this scenario. if cypress test is a must, we will add.

@riodeuno

sbalaji1192 commented on app/client/src/widgets/TableWidget/component/TableUtilities.tsx:
```diff
@@ -632,10 +632,11 @@ export function getDefaultColumnProperties(
   widgetName: string,
   isDerived?: boolean,
 ): ColumnProperties {
+  const id = isNaN(Number(accessor)) ? accessor : `_${accessor}`;
```
Lets extract this into a util function

sbalaji1192 commented on app/client/src/widgets/TableWidget/widget/derived.js:
```diff
@@ -118,10 +118,13 @@ export default {
             .split(separatorRegex)
             .join("_")
             .slice(0, 200);
+          sanitizedKey = _.isNaN(Number(sanitizedKey))
```
here as well.

riodeuno commented on app/client/src/widgets/TableWidget/widget/derived.test.js:
```diff
@@ -1171,6 +1171,63 @@ describe("Validates Derived Properties", () => {
+          demo: "3",
+          demo_1: "1",
+          "test one": "1",
+          "test 3 4 9": "4",
```
Shall we add a few more entries here to test for emojis and special characters?
Do emojis cause this issue we're fixing?

shwetha-ramesh replied:
> Shall we add a few more entries here to test for emojis and special characters? Do emojis cause this issue we're fixing?

@techbhavin  @somangshu 
Tested this PR, numeric column names now gets binded with input widget & also displayed in autosuggest with underscore prefixed.
However wrt comment by @riodeuno are we looking at emojis and special character in this current PR ? Currently column names with special characters are not visible after binding (refer video)
https://loom.com/share/73f5be58dac946e38734aedb960f74bf

sbalaji1192 replied:
@somangshu What should be expected behavior here?

somangshu replied:
WRT to consistency for now between table v1 and v2, we should have a similar behaviour. We should allow all kinds of characters. 

If this is hard fix, we should probably de-scope it and take it up in the table widget v2. Essentially we need to fix this early as we are starting off with v2 in the next quarter and dont want v1 eating up our time

sbalaji1192 replied:
Let's descope it, for now, guessing the effort must be hard to figure out of solution here. Don't wanna spend time solving this, since anyways it's gonna get solved in v2.

sbalaji1192 replied:
@somangshu FYI on the comment above.

somangshu replied:
@sbalaji1192 agreed to this

riodeuno reviewed (dismissed):
@techbhavin Let's move ahead with the cypress test. This is a critical workflow for our table widget. Let's also move forward with a migration. I'm not sure which properties will need migration here. We use the keys in `primaryColumns` `derivedColumns` `columnOrder` `hiddenColumns` etc.

techbhavin commented:
/ok-to-test sha=24499b1

techbhavin commented:
/ok-to-test sha=22c4b8d

techbhavin commented:
/ok-to-test sha=fb8c6ed

techbhavin commented:
/ok-to-test sha=b00436b

techbhavin commented:
/ok-to-test sha=b00436b

shwetha-ramesh commented:
Not taking up column names with special character in this PR.
Moving to Done.

techbhavin commented:
/ok-to-test sha=91bb738

techbhavin commented:
/ok-to-test sha=91bb738

sbalaji1192 commented on app/client/src/utils/migrations/TableWidget.ts:
```diff
@@ -486,3 +489,57 @@ export const migrateTableWidgetIconButtonVariant = (currentDSL: DSLWidget) => {
+export const migrateTableWidgetNumericColumnName = (currentDSL: DSLWidget) => {
+  currentDSL.children = currentDSL.children?.map((child: WidgetProps) => {
+    if (child.type === "TABLE_WIDGET") {
+      child.columnOrder = (child.columnOrder || []).map((co: string) =>
```
make `co` either `column` or `col`.

sbalaji1192 reviewed (approved):
LGTM

techbhavin commented:
/ok-to-test sha=589964b
